pkg/dashboard/manager.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/pingcap/pd/v4/pkg/logutil"
+	"github.com/pingcap/pd/v4/server"
+	"github.com/pingcap/pd/v4/server/cluster"
+)
+
+type electionStatus int
+
+const (
+	stopped electionStatus = iota
+	unknown
+	known
+)
+
+var (
+	// FastCheckInterval is the interval to check dashboard address when it is unknown.
+	FastCheckInterval = time.Second
+	// SlowCheckInterval is the interval to check dashboard address when it is known.
+	SlowCheckInterval = time.Minute
+)
+
+// Manager is used to control dashboard.
+type Manager struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	srv        *server.Server
+	service    *apiserver.Service
+	redirector *Redirector
+}
+
+// NewManager creates a new Manager.
+func NewManager(srv *server.Server, s *apiserver.Service, redirector *Redirector) *Manager {
+	ctx, cancel := context.WithCancel(srv.Context())
+	return &Manager{
+		ctx:        ctx,
+		cancel:     cancel,
+		srv:        srv,
+		service:    s,
+		redirector: redirector,
+	}
+}
+
+func (m *Manager) start() {
+	m.wg.Add(1)
+	go func() {
+		defer logutil.LogPanic()
+		defer m.wg.Done()
+
+		status := unknown
+		for status != stopped {
+			status = m.serviceLoop(status)
+		}
+	}()
+}
+
+func (m *Manager) stop() {
+	m.cancel()
+	m.wg.Wait()
+	log.Info("exit dashboard loop")
+}
+
+func (m *Manager) serviceLoop(lastStatus electionStatus) (nextStatus electionStatus) {
+	var ticker *time.Ticker
+	switch lastStatus {
+	case unknown:
+		ticker = time.NewTicker(FastCheckInterval)
+	case known:
+		ticker = time.NewTicker(SlowCheckInterval)
+	default:
+		panic("unreachable")
+	}
+	defer ticker.Stop()
+
+	nextStatus = lastStatus
+	for nextStatus == lastStatus {
+		select {
+		case <-m.ctx.Done():
+			m.stopService()
+			return stopped
+		case <-ticker.C:
+			nextStatus = m.checkAddress()
+		}
+	}
+
+	return
+}
+
+// checkDashboardAddress checks if the dashboard service needs to change due to dashboard address is changed.
+func (m *Manager) checkAddress() electionStatus {
+	dashboardAddress := m.srv.GetScheduleOption().GetDashboardAddress()
+	switch dashboardAddress {
+	case "auto":
+		if m.srv.GetMember().IsLeader() {
+			m.setNewDashboardAddress()
+		}
+		return unknown
+	case "none":
+		m.redirector.SetAddress("")
+		m.stopService()
+		return known
+	default:
+		if _, err := url.Parse(dashboardAddress); err != nil {
+			log.Error("illegal dashboard address", zap.String("address", dashboardAddress))
+			return unknown
+		}
+		if m.srv.GetMember().IsLeader() && !m.isOnline(dashboardAddress) {
+			m.setNewDashboardAddress()
+			return unknown
+		}
+	}
+
+	m.redirector.SetAddress(dashboardAddress)
+
+	clientUrls := m.srv.GetMemberInfo().GetClientUrls()
+	if len(clientUrls) > 0 && clientUrls[0] == dashboardAddress {
+		m.startService()
+	} else {
+		m.stopService()
+	}
+
+	return known
+}
+
+func (m *Manager) isOnline(addr string) bool {
+	rc := m.srv.GetRaftCluster()
+	if rc == nil || !rc.IsRunning() {
+		return true
```

Comment:
why it is online in this case?

Reply:
At this time, it is impossible to determine whether it is online, so it cannot be handled as offline.